Feature/7929/git info endpoint

LazarenkoDmytro opened the pull request:
#7929

LazarenkoDmytro committed:
Add DTOs for commit information handling

Introduce `CommitInfoDto` as a base class along with `CommitInfoErrorDto` and `CommitInfoSuccessDto` for differentiated error and success responses. These classes will streamline commit-related data transfer and enhance response structure.

LazarenkoDmytro committed:
Add CommitInfoService interface for fetching Git info

Introduce a new interface to retrieve Git commit details, providing a method to fetch the latest commit hash and date.

LazarenkoDmytro committed:
Fix: Conflict resolution between Lombok annotations

The conflict between the @AllArgsConstructor and @NoArgsConstructor annotations has been resolved. These annotations were conflicting because the CommitInfoDto class lacks any fields. Consequently, they attempted to create a duplicate constructor.

LazarenkoDmytro committed:
Add: The implementation of the CommitInfoService interface

Utilized the JGit library to retrieve commit information. Implemented the CommitInfoServiceImpl and conducted comprehensive testing to ensure its functionality.

LazarenkoDmytro committed:
Add CommitInfoController with endpoint and unit tests

Introduced a new REST controller `CommitInfoController` to fetch the latest Git commit hash and date. Added corresponding unit tests in `CommitInfoControllerTest` and updated `HttpStatuses` with `INTERNAL_SERVER_ERROR` for response descriptions.

LazarenkoDmytro committed:
Refactor test imports and assertions for clarity.

Updated imports in test classes to improve readability and maintain consistency with Mockito and JUnit usage. Consolidated and explicitly listed static imports for assertions and Mockito functions. No logic changes were made to the tests themselves.

VasylyshynDmytro commented on service-api/src/main/java/greencity/dto/commitinfo/CommitInfoDto.java:
```diff
@@ -0,0 +1,12 @@
+@Data
+@NoArgsConstructor
+public abstract class CommitInfoDto {
+}
```
It could be an interface.

VasylyshynDmytro commented on service-api/src/main/java/greencity/dto/commitinfo/CommitInfoErrorDto.java:
```diff
@@ -0,0 +1,17 @@
+@EqualsAndHashCode(callSuper = true)
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
```
annotation
@Data
@NoArgsConstructor already present in abstract class CommitInfo you shouldn't duplicate it(if you want to use abstract class)

VasylyshynDmytro commented on service-api/src/main/java/greencity/dto/commitinfo/CommitInfoSuccessDto.java:
```diff
@@ -0,0 +1,18 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class CommitInfoSuccessDto extends CommitInfoDto {
```
the same situation like we have at CommitInfoErrorDto

VasylyshynDmytro commented on service/src/main/java/greencity/service/CommitInfoServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+            return new CommitInfoSuccessDto(latestCommitHash, latestCommitDate);
+        } catch (IOException e) {
+            return new CommitInfoErrorDto("Failed to fetch commit info due to I/O error: " + e.getMessage());
+        }
```
maybe add this String in to ErrorMessage
"Failed to fetch commit info due to I/O error: "

VasylyshynDmytro commented on service/src/main/java/greencity/service/CommitInfoServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+                .build();
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed to initialize repository", e);
+        }
```
maybe add this String in to ErrorMessage
"Failed to initialize repository"

VasylyshynDmytro commented on service/src/test/java/greencity/service/CommitInfoServiceImplTest.java:
```diff
@@ -0,0 +1,148 @@
+    private PersonIdent personIdent;
+
+    private static final String COMMIT_HASH = "abc123";
+
```
add it to consts

LazarenkoDmytro replied:
It is just a test var - no need to declare it in another file

VasylyshynDmytro commented on core/src/main/java/greencity/controller/CommitInfoController.java:
```diff
@@ -0,0 +1,61 @@
+                value = """
+                    {
+                        "error": "Failed to fetch commit info: Repository not found"
+                    }""")))
```
add this to constant
"error": "Failed to fetch commit info: Repository not found"

VasylyshynDmytro commented on service/src/test/java/greencity/service/CommitInfoServiceImplTest.java:
```diff
@@ -0,0 +1,148 @@
+                when(builderMock.readEnvironment()).thenReturn(builderMock);
+                when(builderMock.findGitDir()).thenReturn(builderMock);
+                when(builderMock.build()).thenThrow(new IOException("Repository not found"));
+            })) {
```
maybe you should add it to constant

LazarenkoDmytro replied:
Probably no because it is used once and will not be changed

RomaMocherniuk commented on service-api/src/main/java/greencity/dto/commitinfo/CommitInfoDto.java:
```diff
@@ -0,0 +1,12 @@
+ */
+@Data
+@NoArgsConstructor
+public abstract class CommitInfoDto {
```
can be an interface

RomaMocherniuk replied:
Or itis  better to create a single class with the next fields
```
 private String commitHash;
 private String commitDate;
 private String error;
```
And in the controller check if the error is present

LazarenkoDmytro replied:
1) The interface requirement is satisfactory.
2) It would be advantageous to have two types of responses. This is elegantly implemented through polymorphism. It is somewhat unappealing to have null in the error field.

RomaMocherniuk commented on service/src/main/java/greencity/service/CommitInfoServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+                .findGitDir()
+                .build();
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed to initialize repository", e);
```
Do we need to fail the application load if the `.git` is not found?

LazarenkoDmytro replied:
Ok, it will be handled

RomaMocherniuk commented on core/src/main/java/greencity/controller/CommitInfoController.java:
```diff
@@ -0,0 +1,61 @@
+    public ResponseEntity<CommitInfoDto> getCommitInfo() {
+        CommitInfoDto commitInfo = commitInfoService.getLatestCommitInfo();
+        if (commitInfo instanceof CommitInfoErrorDto) {
+            return ResponseEntity.internalServerError().body(commitInfo);
```
why do you return an internal server error?
it is a bad practice to return 500 response, its better to return `NOT_FOUND`

LazarenkoDmytro replied:
Will be replaced

RomaMocherniuk commented on service/src/main/java/greencity/service/CommitInfoServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+public class CommitInfoServiceImpl implements CommitInfoService {
+    private Repository repository;
+
+    @PostConstruct
```
why do you need `@PostConstruct` here?

LazarenkoDmytro replied:
Because constructing the Repository can throw an IOException. That's why I do not want to place it inside the constructor

RomaMocherniuk commented on service/src/main/java/greencity/service/CommitInfoServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public CommitInfoDto getLatestCommitInfo() {
+        try (RevWalk revWalk = new RevWalk(repository)) {
+            RevCommit latestCommit = revWalk.parseCommit(repository.resolve("HEAD"));
```
please move `HEAD` to the constant

LazarenkoDmytro replied:
Ok

LazarenkoDmytro committed:
Refactor CommitInfoDto hierarchy to use interface.

Replaced the abstract class CommitInfoDto with an interface for improved flexibility and clarity. Updated related DTOs (CommitInfoErrorDto and CommitInfoSuccessDto) to implement the interface instead of extending the abstract class. This aligns with better design principles and reduces unnecessary coupling.

LazarenkoDmytro committed:
Refactor commit reference to use a constant.

Replaced the hardcoded "HEAD" with a static constant COMMIT_REF for better readability and maintainability. This ensures consistency in resolving the latest commit reference across the service.

LazarenkoDmytro committed:
Refactor repository initialization in CommitInfoService

Replaced @PostConstruct initialization with constructor-based logic to improve clarity and simplify error handling. Added logging for missing `.git` directory and updated related tests to handle uninitialized repository scenarios. Enhanced `getLatestCommitInfo` with proper error handling when the repository is null.

LazarenkoDmytro committed:
Update error status from 500 to 404 for commit info retrieval

Changed the HTTP response code and description for cases where commit info is absent, aligning with proper REST conventions. Updated tests, API documentation, and removed unused constant `INTERNAL_SERVER_ERROR` accordingly.

KizerovDmitriy commented on core/src/main/java/greencity/controller/CommitInfoController.java:
```diff
@@ -0,0 +1,62 @@
+    @GetMapping
+    public ResponseEntity<CommitInfoDto> getCommitInfo() {
+        CommitInfoDto commitInfo = commitInfoService.getLatestCommitInfo();
+        if (commitInfo instanceof CommitInfoErrorDto) {
```
Maybe better to move this check to service layer?

LazarenkoDmytro replied:
It seems that this behavior can be only here. I return different statuses (ok and not_found) for different DTOs

KizerovDmitriy commented on service/src/main/java/greencity/service/CommitInfoServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+
+            return new CommitInfoSuccessDto(latestCommitHash, latestCommitDate);
+        } catch (IOException e) {
+            return new CommitInfoErrorDto("Failed to fetch commit info due to I/O error: " + e.getMessage());
```
Please add message to ErrorMessage utility class.

LazarenkoDmytro replied:
OK

LazarenkoDmytro committed:
Replaced the polymorphism on DTOs with the single DTO

LazarenkoDmytro committed:
Adjusted the Service and the Implementation to follow the new DTO structure. Also extracted the error messages in a separate file

LazarenkoDmytro committed:
Adjusted the Controller to follow the new DTO structure. Also extracted the error messages in a separate file. Set the default timezone in the JSON for the 404 error

LazarenkoDmytro committed:
Allowed all the users to access the endpoint

LazarenkoDmytro committed:
Fixed the 120 characters requirement violation (just deleted the useless text for the example)
